src/expire-tiles.cpp

```diff
@@ -134,6 +135,46 @@ void expire_tiles::coords_to_tile(double lon, double lat, double *tilex,
     *tiley = map_width * (0.5 - c.y() / EARTH_CIRCUMFERENCE);
 }
 
+void expire_tiles::from_point_list(geom::point_list_t const &list)
+{
+    for_each_segment(list, [&](geom::point_t a, geom::point_t b) {
+        from_line(a.x(), a.y(), b.x(), b.y());
+    });
+}
+
+void expire_tiles::from_geometry(geom::geometry_t const &geom, osmid_t osm_id)
+{
+    if (geom.srid() != 3857) {
+        return;
+    }
+
+    if (geom.is_point()) {
+        auto const box = geom::envelope(geom);
+        from_bbox(box);
+    } else if (geom.is_linestring()) {
+        from_point_list(geom.get<geom::linestring_t>());
+    } else if (geom.is_multilinestring()) {
+        for (auto const &list : geom.get<geom::multilinestring_t>()) {
+            from_point_list(list);
+        }
+    } else if (geom.is_polygon() || geom.is_multipolygon()) {
+        auto const box = geom::envelope(geom);
+        if (from_bbox(box)) {
+            /* Bounding box too big - just expire tiles on the line */
+            log_debug("Large polygon ({:.0f} x {:.0f} metres, OSM ID {})"
+                      " - only expiring perimeter",
+                      box.max_x() - box.min_x(), box.max_y() - box.min_y(), osm_id);
+            if (geom.is_polygon()) {
+                from_point_list(geom.get<geom::polygon_t>().outer());
+            } else if (geom.is_multipolygon()) {
+                for (auto const &polygon : geom.get<geom::multipolygon_t>()) {
+                    from_point_list(polygon.outer());
```

Comment:
The original code expires all rings here.

Reply:
Fixed and force-pushed.